test/ASTGen/keypath.swift

```diff
@@ -0,0 +1,243 @@
+  subscript(optSub: OptSub) -> A? { get { return A() } set { } }
+  subscript(nonHashableSub: NonHashableSub) -> A { get { return A() } set { } }
+  subscript(a: Sub, b: Sub) -> A { get { return A() } set { } }
+  subscript(a: Sub, b: NonHashableSub) -> A { get { return A() } set { } }
+
+  var nonMutatingProperty: B {
+    get { fatalError() }
+    nonmutating set { fatalError() }
+  }
+}
+
+struct A: Hashable {
+  init() { fatalError() }
+
+  var property: Prop
+  var optProperty: Prop?
+  let optLetProperty: Prop?
+
+  subscript(sub: Sub) -> A { get { return self } set { } }
+
+  static func ==(_: A, _: A) -> Bool { fatalError() }
+  func hash(into hasher: inout Hasher) { fatalError() }
+}
+struct B {}
+struct C<T> { // expected-note 2 {{'T' declared as parameter to type 'C'}}
+  var value: T
+  subscript() -> T { get { return value } }
+  subscript(sub: Sub) -> T { get { return value } set { } }
+  subscript<U: Hashable>(sub: U) -> U { get { return sub } set { } }
+  subscript<X>(noHashableConstraint sub: X) -> X { get { return sub } set { } }
+}
+
+/* FIXME: @available is not implmented in ASTGen.
+struct Unavailable {
+  @available(*, unavailable)
+  var unavailableProperty: Int { 0 }
+  // xpected-note@-1 {{'unavailableProperty' has been explicitly marked unavailable here}}
+
+  @available(*, unavailable)
+  subscript(x: Sub) -> Int { get { } set { } }
+  // xpected-note@-1 {{'subscript(_:)' has been explicitly marked unavailable here}}
+}
+
+struct Deprecated {
+  @available(*, deprecated)
+  var deprecatedProperty: Int
+
+  @available(*, deprecated)
+  subscript(x: Sub) -> Int { get { } set { } }
+}
+
+@available(*, deprecated)
+func getDeprecatedSub() -> Sub {
+  return Sub()
+}
+*/
+
+extension Array where Element == A {
+  var property: Prop { fatalError() }
+}
+
+protocol P { var member: String { get } }
+extension B : P { var member : String { return "Member Value" } }
+
+struct Exactly<T> {}
+
+func expect<T>(_ x: inout T, toHaveType _: Exactly<T>.Type) {}
+
+func testKeyPath(sub: Sub, optSub: OptSub,
+                 nonHashableSub: NonHashableSub, x: Int) {
+  var a = \A.property
+  expect(&a, toHaveType: Exactly<WritableKeyPath<A, Prop>>.self)
+
+  var b = \A.[sub]
+  expect(&b, toHaveType: Exactly<WritableKeyPath<A, A>>.self)
+
+  var c = \A.[sub].property
+  expect(&c, toHaveType: Exactly<WritableKeyPath<A, Prop>>.self)
+
+  var d = \A.optProperty?
+  expect(&d, toHaveType: Exactly<KeyPath<A, Prop?>>.self)
+
+  var e = \A.optProperty?[sub]
+  expect(&e, toHaveType: Exactly<KeyPath<A, A?>>.self)
+
+  var f = \A.optProperty!
+  expect(&f, toHaveType: Exactly<WritableKeyPath<A, Prop>>.self)
+
+  var g = \A.property[optSub]?.optProperty![sub]
+  expect(&g, toHaveType: Exactly<KeyPath<A, A?>>.self)
+
+  var h = \[A].property
+  expect(&h, toHaveType: Exactly<KeyPath<[A], Prop>>.self)
+
+  var i = \[A].property.nonMutatingProperty
+  expect(&i, toHaveType: Exactly<ReferenceWritableKeyPath<[A], B>>.self)
+
+  var j = \[A].[x]
```

Comment:
Amusingly you can also do `\[A][x]` 😄

Reply:
Hm I thought we require a dot if the first "path" part is a subscript. But apparently not 🤔

Reply:
FWIW this cases are pulled from `test/expr/unary/keypath/keypath.swift` And I will add more test cases from it when I revisit here for adding AST diagnostics (for e.g. empty components)